Sources/AsyncKit/ConnectionPool.swift

```diff
@@ -216,23 +272,40 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
     /// Connection pools must be closed before they deinitialize.
     ///
     /// - returns: A future indicating close completion.
-    public func close() -> EventLoopFuture<Void> {
-        self.isClosed = true
-        return self.available.map { $0.close() }.flatten(on: self.eventLoop).map {
-            // inform any waiters that they will never be receiving a connection
-            while let waiter = self.waiters.popFirst() {
-                waiter.fail(ConnectionPoolError.closed)
+    public func shutdown() {
+        // synchronize access to closing
+        self.lock.lock()
+        defer { self.lock.unlock() }
+
+        // check to make sure we aren't double closing
+        guard !self.didShutdown else {
+            return
+        }
+        self.didShutdown = true
+
+        // no locks needed as this can only happen once
+        for available in self.available {
+            do {
+                try available.close().wait()
+            } catch {
+                #warning("TODO: use logger")
+                print("Could not close connection: \(error)")
             }
-            
-            // reset any variables to free up memory
-            self.available = []
-            self.activeConnections = 0
         }
+
+        // inform any waiters that they will never be receiving a connection
+        while let waiter = self.waiters.popFirst() {
+            waiter.fail(ConnectionPoolError.shutdown)
+        }
+            
+        // reset any variables to free up memory
+        self.available = []
+        self.activeConnections.store(0)
```

Comment:
FYI, I think `activeConnections` can become negative when connections are released (and subsequently closed) _after_ `shutdown` has been called. You could do `self.activeConnections -= available.count; available = []` instead.